docs/mat.1.in
.TH mat "1" "2017-02-08" "@VERSION@" "mattools manual"
.SH NAME
mattools \- utilities for distance matrices
.SH SYNOPSIS
.B mat \fICOMMAND\fR
[\fIOPTIONS\fR] \fIFILES\fR...
.SH DESCRIPTION
.TP
The \fBmattools\fR are a set of utilities for the manipulation, formatting and comparison of distance matrices in PHYLIP format. The following commands are available. See below for a list of options.

.TP
mat \fBcompare\fR [\fIOPTIONS\fR] \fIFILES\fR...
Compute the distance between two matrices.
.TP
mat \fBformat\fR [\fIOPTIONS\fR] \fIFILES\fR...
Interpret the input as a distance matrix and provide a properly formatted output.
.TP
mat \fBgrep\fR \fIPATTERN\fR [\fIOPTIONS\fR] \fIFILES\fR...
Print only the submatrix for names matching a the given regular expression. The \fIPATTERN\fR is assumed to be in ECMAScript (JavaScript) syntax.
.TP
mat \fBnj\fR [\fIOPTIONS\fR] \fIFILES\fR...
Build a tree by neighbor joining and outputs it in NEWICK format. Also computes support values via quartet analysis.

.LP
All commands read PHYLIP distance matrices from the given files. If no file names are supplied, \fBmat\fR reads from \fIstdin\fR instead.


.SH PHYLIP DISTANCE MATRICES

The PHYLIP distance matrix is a terrible data format but commonly used. On its first line it states the dimension of the matrix \fIN\fR. Then follow \fIN\fR rows, each starting with the name of a sequence. If necessary, the name is padded with spaces to fill ten characters.

    3
    S0         0.0000 0.0957 0.0975
    S1         0.0957 0.0000 0.1822
    S2         0.0975 0.1822 0.0000

Behind each name follow up to \fIN\fR entries of the matrix. Each entry is separated by a blank space. Missing input values are set to zero.

There also exists a lower triangular format, with or without main diagonal. This is accepted as input, but the mattools will always output full square matrices.


.SH GENERAL OPTIONS
.TP
\fB\--version\fR
Outputs version information.
.TP
\fB\--help\fR
Print general help.


.SH COMPARE OPTIONS
.TP
\fB\--delta2\fR
Use a Fitch-Margolish like measure.
.TP
\fB\--help\fR
Print help for compare command.
.TP
\fB\--rel\fR
Compute the average relative dissimilarity.

.SH FORMAT OPTIONS
.TP
\fB\-f\fR, \fB\--fix\fR
With this option, small errors are fixed automatically. These include asymmetric values, non-zero values on the main diagonal, and negative values.
.TP
\fB\--format=\fR\fISTRING\fR
This option can be used to change the way each value of the distance matrix is printed. The \fISTRING\fR is interpreted as a C-style \fBprintf\fR(3) format string. The only given argument is a \fIdouble\fR. The default value of \fISTRING\fR is \fB%1.4e\fR.
.TP
\fB\--precision=\fR\fIPRECISION\fR
Floats cannot be compared equal, thus two values with a relative error of \fIPRECISION\fR are considered the same. The default value for \fIPRECISION\fR is \fB0.05\fR.
.TP
\fB\--separator=\fR\fISEP\fR
Use the \fISEP\fR character to separate columns in the output. The default value for \fISEP\fR is a blank space i.e. ' '. \fISEP\fR can also interpret C-style escape codes such as \fB\\t\fR.
.TP
\fB\-s\fR, \fB\--sort\fR
Rearrange the matrix by sorting the names lexicographically. Useful for canonizing output.
.TP
\fB--truncate-names\fR
Truncate names to ten characters.
.TP
\fB\-v\fR, \fB\--validate\fR
Validate the input for correctness: Checks for double names, zeros beyond the main diagonal, NaNs, and also verifies triangle inequality. Implicitly applies \fB--fix\fR before validation.
.TP
\fB-h\fR, \fB\--help\fR
Print help for format command.


.SH GREP OPTIONS
.TP
\fB\-v\fR, \fB\--invert-match\fR
Keep all names that do not match the pattern.
.TP
\fB-h\fR, \fB\--help\fR
Print help for grep command.


.SH NEIGHBOR JOINING OPTIONS
.TP
\fB-h\fR, \fB\--help\fR
Print help for neighbor joining command.
.TP
\fB--no-support\fR
Do not compute support values.


.SH COPYRIGHT
Copyright \(co 2017-2019 Fabian Klötzl
License GPLv3+: GNU GPL version 3 or later.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
The full license text is available at <http://gnu.org/licenses/gpl.html>.
.PP
.SH BUGS
.SS Reporting Bugs
Please report bugs to <kloetzl@evolbio.mpg.de> or at <https://github.com/kloetzl/mattools>.
.SS


src/mat.cxx
/*
 * Copyright (C) 2017  Fabian Klötzl
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <err.h>
#include <stdio.h>
#include <string>

int mat_combine(int, char **);
int mat_compare(int, char **);
int mat_diff(int, char **);
int mat_grep(int, char **);
int mat_nj(int, char **);
int mat_format(int, char **);
int mat_mantel(int, char **);
static void usage(int status);
static void version();

/** @brief The main function of the mat tools. It dispatches into the sub
 * commands.
 *
 * @param argc - argc.
 * @param argv - argv.
 * @returns 0 iff successful.
 */
int main(int argc, char *argv[])
{
	if (argc < 2) {
		usage(EXIT_FAILURE);
	}

	auto first_arg = std::string{argv[1]};
	if (first_arg == "--version") {
		version();
		exit(EXIT_SUCCESS);
	}

	if (first_arg == "--help") {
		usage(EXIT_SUCCESS);
	}

	// strip binary
	argc -= 1, argv += 1;

	auto command = first_arg;
	if (command == "combine") {
		return mat_combine(argc, argv);
	}

	if (command == "compare") {
		return mat_compare(argc, argv);
	}

	if (command == "diff") {
		return mat_diff(argc, argv);
	}

	if (command == "grep") {
		return mat_grep(argc, argv);
	}

	if (command == "nj") {
		return mat_nj(argc, argv);
	}

	if (command == "format") {
		return mat_format(argc, argv);
	}

	if (command == "mantel") {
		return mat_mantel(argc, argv);
	}

	warnx("unknown command '%s'.", command.c_str());
	usage(EXIT_FAILURE);

	return 1;
}

static void usage(int status)
{
	static const char str[] = {
		"usage: mat [--version] [--help] <command> [<args>]\n\n"
		"The available commands are:\n"
		" compare     Compute the distance between two matrices\n"
		" format      Format the distance matrix\n"
		" grep        Print submatrix for names matching a pattern\n"
		" nj          Convert to a tree by neighbor joining\n"
		"\n"
		"Use 'mat <command> --help' to get guidance on the usage of a "
		"command.\n"};

	fprintf(status == EXIT_SUCCESS ? stdout : stderr, str);
	exit(status);
}

static void version()
{
	static const char str[] = {PACKAGE_STRING "\n"};
	printf(str);
}
